apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,244 @@
+	// teams within the listed GitHub organizations. Additional login rules or group filtering can optionally be
+	// provided as policy expression on any Pinniped Supervisor FederationDomain that includes this IDP.
+	//
+	// The configured GitHub App or GitHub OAuth App must be allowed to see membership in the listed organizations,
+	// otherwise Pinniped will not be aware that the user belongs to the listed organization or any teams
+	// within that organization.
+	//
+	// If no organizations are listed, you must set gitHubOrganizationLoginPolicy: AllOrganizationsAllowed
+	//
+	// +kubebuilder:validation:MaxItems=64
+	// +listType=set
+	// +optional
+	AllowedOrganizations []string `json:"allowedOrganizations,omitempty"`
+
+	// OrganizationLoginPolicy must be set to "AllOrganizationsAllowed" if allowedOrganizations is empty.
+	//
+	// This field only exists to ensure that Pinniped administrators are aware that an empty list of
+	// allowedOrganizations means all GitHub users are allowed to log in.
+	//
+	// +kubebuilder:default=AllowedOrganizationsOnly
+	// +kubebuilder:validation:Enum=AllowedOrganizationsOnly;AllOrganizationsAllowed
+	// +optional
+	OrganizationLoginPolicy *GitHubOrganizationLoginPolicy `json:"organizationLoginPolicy"`
+
+	// Client identifies the secret with credentials for a GitHub App or GitHub OAuth2 App (a GitHub client).
+	Client GitHubClientSpec `json:"client"`
+}
+
+// GitHubIdentityProvider describes the configuration of an upstream GitHub identity provider.
+// This upstream provider can be configured with either a GitHub App or a GitHub OAuth2 App.
+//
+// Right now, only web-based logins are supported, for both the pinniped-cli client and clients configured
+// as OIDCClients.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories=pinniped;pinniped-idp;pinniped-idps
+// +kubebuilder:printcolumn:name="Host",type=string,JSONPath=`.spec.gitHubAPI.host`
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.phase`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:subresource:status
+type GitHubIdentityProvider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec for configuring the identity provider.
+	// +kubebuilder:validation:XValidation:message="organizationLoginPolicy must be 'AllowedOrganizationsOnly' if allowedOrganizations are listed, or 'AllOrganizationsAllowed' if no allowedOrganizations are listed",rule="(self.organizationLoginPolicy == 'AllowedOrganizationsOnly' && has(self.allowedOrganizations) && size(self.allowedOrganizations) > 0) || (self.organizationLoginPolicy == 'AllOrganizationsAllowed' && (!has(self.allowedOrganizations) || size(self.allowedOrganizations) < 1))"
+	Spec GitHubIdentityProviderSpec `json:"spec"`
```

Comment:
Would this user-facing error message be easier to understand if it were split into two validations with errors like:

- the enum field must have value X when the list is empty
- the enum field must have value Y when the list is non-empty

?

Reply:
Yes, this has been done.